### ISSUE TITLE
Dockerizing issue

### ISSUE THREAD

**Issue by richardforth** [richardforth/apache2buddy#329]
I have an issue dockerizing apache2buddy, Im getting hung up on pmap outputs looking different than on a regular server, and its causing a roadblock, on a normal server Im looking for this line :

```bash
# pmap -d 1108  | tail -1
mapped: 249856K    writeable/private: 2464K    shared: 76K
```

I havent got any example output from docker, but its the same across debian:latest, centos:lastest centos:7 etc,  I know its a docker thing and Im missing something.

Once I've got apache2buddy  containerised I can run faster tests on staging env

**Comment by richardforth** [comment]
Going to work on creating a Dockerfile for repeatable testing of this issue and paste it in here later. Im new to Docker. But this is much needed as I dont have a CI/CD pipeline for apache2buddy and since it actually costs me money I need to do it all on the cheap, getting working docker containers will help work towards something that vaguely resembles CI/CD and at least partially automates testing on multiple OS flavours.

**Comment by richardforth** [comment]
My Dockerfile:

```bash
FROM centos:7
RUN yum install -y python3
RUN yum install -y perl
RUN yum install -y which
RUN yum install -y curl
RUN yum install -y net-tools
RUN yum install -y procps
RUN yum install -y httpd
RUN yum install -y httpd-tools
RUN /usr/sbin/httpd -k start && curl -sk https://raw.githubusercontent.com/richardforth/apache2buddy/staging/apache2buddy.pl | perl 
```

**Comment by richardforth** [comment]
```bash
[root@8fa6a38eb5f8 /]# curl -sk https://raw.githubusercontent.com/richardforth/apache2buddy/staging/apache2buddy.pl | perl
#################################################
apache2buddy.pl report for 8fa6a38eb5f8 (x.x.x.x)
#################################################
[ OK ] This script is being run as root.
[ OK ] The utility 'pmap' exists and is available for use: /usr/bin/pmap
[ OK ] The utility 'netstat' exists and is available for use: /usr/bin/netstat
which: no php in (/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/root/bin)
[ @@ ] Unable to locate the PHP binary. PHP specific checks will be skipped.
[ OK ] The utility 'apachectl' exists and is available for use: /usr/sbin/apachectl
[ OK ] The port (port 80) is a valid port.
[ -- ] We are attempting to discover the operating system type and version number ...
[ OK ] The 'python' binary exists and is available for use: /usr/bin/python
[ -- ] Distro: CentOS Linux
[ -- ] Version: 7.8.2003
[ -- ] Codename: Core
[ OK ] This distro is supported by apache2buddy.pl.
[ OK ] This distro version is supported by apache2buddy.pl.
[ -- ] Hostname: 8fa6a38eb5f8
[ -- ] Primary IP: x.x.x.x
[ -- ] We are checking the service running on port 80...
[ -- ] The process listening on port 80 is /usr/sbin/httpd
[ -- ] The process running on port 80 is Apache/2.4.6 (CentOS).
[ -- ] The full path to the Apache config file is: /etc/httpd/conf/httpd.conf
[ -- ] Apache is using prefork model.
[ -- ] pidfile setting is CONFIG NOT FOUND.
[ -- ] Actual pidfile is /run/httpd/httpd.pid.
[ -- ] Parent PID: 21.
[ OK ] Memory usage of parent PID is less than 50MB: 1776 Kilobytes.
[ -- ] Apache has been running 0d 0h 02m 04s.
[ !! ] *** LOW UPTIME ***.
[ @@ ] The following recommendations may be misleading - apache has been restarted within the last 24 hours.
[ -- ] Your server has 1991 MB of PHYSICAL memory.
[ >> ] ServerLimit directive not found, assuming default values.
[ -- ] Your ServerLimit setting is 256.
[ >> ] MaxRequestWorkers directive not found, assuming default values.
[ -- ] Your MaxRequestWorkers setting is 256.
[ OK ] Current Apache Process Count is 6, including the parent PID.
[ -- ] Number of vhosts detected: 0.
[ OK ] Current Apache vHost Count is less than maxrequestworkers.
[ @@ ] vHost Count works only when we have NameVirtualHosting enabled, check config manually, they may only have the default vhost.
[ >> ] MaxRequestsPerChild directive not found.
[ >> ] MaxConnectionsPerChild directive not found.
[ -- ] This server is NOT running Plesk.
[ -- ] This server is NOT running cPanel.
[ -- ] This server is NOT running Virtualmin.
[ OK ] No additional services were detected.
[ OK ] No large log files were found in /var/log/httpd.
[ OK ] MaxClients has not been hit recently.
[ >> ] Apache only logs maxclients/maxrequestworkers hits once in a lifetime, if no restart has happened this event may have been rotated away.
[ >> ] As a backup check, please compare number of running apache processes (minus 1 for parent) against maxclients/maxrequestworkers.
[ >> ] For more information see https://github.com/apache/httpd/blob/0b61edca6cdda2737aa1d84a4526c5f9d2e23a8c/server/mpm/prefork/prefork.c#L809
[ OK ] No package updates found.
[ -- ] httpd is currently using 35.58 MB of memory.
Use of uninitialized value $proc_mem_usages[0] in division (/) at - line 900 (#1)
    (W uninitialized) An undefined value was used as if it were already
    defined.  It was interpreted as a "" or a 0, but maybe it was a mistake.
    To suppress this warning assign a defined value to your variables.
    
    To help you figure out what was undefined, perl will try to tell you
    the name of the variable (if any) that was undefined.  In some cases
    it cannot do this, so it also tells you what operation you used the
    undefined value in.  Note, however, that perl optimizes your program
    anid the operation displayed in the warning may not necessarily appear
    literally in your program.  For example, "that $foo" is usually
    optimized into "that " . $foo, and the warning will refer to the
    concatenation (.) operator, even though there is no . in
    your program.
    
Use of uninitialized value $proc_mem_usages[0] in division (/) at - line 906 (#1)
Use of uninitialized value $count in division (/) at - line 920 (#1)

Illegal division by zero at - line 920 (#2)
    (F) You tried to divide a number by 0.  Either something was wrong in
    your logic, or you need to put a conditional in to guard against
    meaningless input.
    
Uncaught exception from user code:
	Illegal division by zero at - line 920.
	main::get_memory_usage('/usr/sbin/httpd', 'apache', 'average') called at - line 2668
[root@8fa6a38eb5f8 /]# 
```

**Comment by richardforth** [comment]
My problem seems to be that in docker only the Parent PID shows writeable/private line, all the child pids dont have it, but it needs this for each child to calculate smallest, average, and largest process sizes:

```bash
[root@8fa6a38eb5f8 /]# pmap -d 21 | grep write
00007fe29dd01000      60 r-x-- 0000000000000000 0fe:00001 mod_rewrite.so
00007fe29dd10000    2048 ----- 000000000000f000 0fe:00001 mod_rewrite.so
00007fe29df10000       4 r---- 000000000000f000 0fe:00001 mod_rewrite.so
00007fe29df11000       4 rw--- 0000000000010000 0fe:00001 mod_rewrite.so
mapped: 224092K    writeable/private: 1776K    shared: 76K
[root@8fa6a38eb5f8 /]# pmap -d 22 | grep write
[root@8fa6a38eb5f8 /]# pmap -d 23 | grep write
[root@8fa6a38eb5f8 /]# pmap -d 24 | grep write
[root@8fa6a38eb5f8 /]# pmap -d 25 | grep write
[root@8fa6a38eb5f8 /]# pmap -d 26 | grep write
[root@8fa6a38eb5f8 /]# 
```

in this case only PID 21 in docker has the writeable/private line. on a normal server, all the child pids have this also.

**Comment by richardforth** [comment]
I get the same issue on different builds, like debian:latest and centos:latest too.

**Comment by richardforth** [comment]
If you need a line number in the apache2buddy source code, its the block starting at line 869:

```perl
        # figure out how much memory each process is using
	foreach (@pids) {
		chomp($_);
		# pmap -d is used to determine the memory usage for the 
		# individual processes
		my ($distro, $version, $codename) = get_os_platform();
		#output of 'pmap' is different depending on distro!
		my $pid_mem_usage;
		if (ucfirst($distro) eq "SUSE Linux Enterprise Server" ) {
			$pid_mem_usage = `LANGUAGE=en_GB.UTF-8 pmap -d $_ | egrep "writable-private" | awk \'{ print \$1 }\'`;
		} else {
			$pid_mem_usage = `LANGUAGE=en_GB.UTF-8 pmap -d $_ | egrep "writeable/private" | awk \'{ print \$4 }\'`;
		}
		$pid_mem_usage =~ s/K//;
		chomp($pid_mem_usage);

		print "VERBOSE: Memory usage by PID ".$_." is ".$pid_mem_usage."K\n" if $main::VERBOSE;
		
		# on a busy system, the grep output will return the pid for the
		# grep process itself, which will be gone by the time we get 
		# around to running pmap
		if ( $pid_mem_usage ne "" ) {
			push(@proc_mem_usages, $pid_mem_usage);
		}
	}
```

**Comment by richardforth** [comment]
This is not a problem on a real OS, its only an issue in Docker.

**Comment by richardforth** [comment]
found interesting reference:

https://github.com/elastic/elasticsearch-docker/issues/208#issuecomment-438121523

TLDR; "I don't know why yet, but adding the SYS_PTRACE capability restores the desired behaviour."

```bash
--cap-add SYS_PTRACE
```

Hopefully this solves my problem too.... will need to try it, and find out how to "add capabilities", in my Dockerfile

**Comment by richardforth** [comment]
Ok so apparrently...

https://stackoverflow.com/questions/50819439/assigning-additional-capabilities-using-a-docker-file

they are just docker run options, to add capabilities, will give it a whirl later

**Comment by richardforth** [comment]
This worked: 

```bash
$ docker run --cap-add SYS_PTRACE -ti 1c6790507420 /bin/bash
[root@cef6cc345c40 /]# ps auxfww
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.3  0.1  11840  2892 pts/0    Ss   12:15   0:00 /bin/bash
root        14  0.0  0.1  51764  3368 pts/0    R+   12:15   0:00 ps auxfww
[root@cef6cc345c40 /]# /usr/sbin/httpd -k start &
[1] 15
[root@cef6cc345c40 /]# AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 172.17.0.2. Set the 'ServerName' directive globally to suppress this message

[1]+  Done                    /usr/sbin/httpd -k start
[root@cef6cc345c40 /]# 
[root@cef6cc345c40 /]# curl -sk https://raw.githubusercontent.com/richardforth/apache2buddy/staging/apache2buddy.pl | perl 
#######################################################
apache2buddy.pl report for cef6cc345c40 (81.147.138.39)
#######################################################
[ OK ] This script is being run as root.
[ OK ] The utility 'pmap' exists and is available for use: /usr/bin/pmap
[ OK ] The utility 'netstat' exists and is available for use: /usr/bin/netstat
which: no php in (/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin)
[ @@ ] Unable to locate the PHP binary. PHP specific checks will be skipped.
[ OK ] The utility 'apachectl' exists and is available for use: /usr/sbin/apachectl
[ OK ] The port (port 80) is a valid port.
[ -- ] We are attempting to discover the operating system type and version number ...
[ OK ] The 'python' binary exists and is available for use: /usr/bin/python
[ -- ] Distro: CentOS Linux
[ -- ] Version: 7.8.2003
[ -- ] Codename: Core
[ OK ] This distro is supported by apache2buddy.pl.
[ OK ] This distro version is supported by apache2buddy.pl.
[ -- ] Hostname: cef6cc345c40
[ -- ] Primary IP: 81.147.138.39
[ -- ] We are checking the service running on port 80...
[ -- ] The process listening on port 80 is /usr/sbin/httpd
[ -- ] The process running on port 80 is Apache/2.4.6 (CentOS).
[ -- ] The full path to the Apache config file is: /etc/httpd/conf/httpd.conf
[ -- ] Apache is using prefork model.
[ -- ] pidfile setting is CONFIG NOT FOUND.
[ -- ] Actual pidfile is /run/httpd/httpd.pid.
[ -- ] Parent PID: 16.
[ OK ] Memory usage of parent PID is less than 50MB: 1776 Kilobytes.
[ -- ] Apache has been running 0d 0h 00m 55s.
[ !! ] *** LOW UPTIME ***.
[ @@ ] The following recommendations may be misleading - apache has been restarted within the last 24 hours.
[ -- ] Your server has 1991 MB of PHYSICAL memory.
[ >> ] ServerLimit directive not found, assuming default values.
[ -- ] Your ServerLimit setting is 256.
[ >> ] MaxRequestWorkers directive not found, assuming default values.
[ -- ] Your MaxRequestWorkers setting is 256.
[ OK ] Current Apache Process Count is 6, including the parent PID.
[ -- ] Number of vhosts detected: 0.
[ OK ] Current Apache vHost Count is less than maxrequestworkers.
[ @@ ] vHost Count works only when we have NameVirtualHosting enabled, check config manually, they may only have the default vhost.
[ >> ] MaxRequestsPerChild directive not found.
[ >> ] MaxConnectionsPerChild directive not found.
[ -- ] This server is NOT running Plesk.
[ -- ] This server is NOT running cPanel.
[ -- ] This server is NOT running Virtualmin.
[ OK ] No additional services were detected.
[ OK ] No large log files were found in /var/log/httpd.
[ OK ] MaxClients has not been hit recently.
[ >> ] Apache only logs maxclients/maxrequestworkers hits once in a lifetime, if no restart has happened this event may have been rotated away.
[ >> ] As a backup check, please compare number of running apache processes (minus 1 for parent) against maxclients/maxrequestworkers.
[ >> ] For more information see https://github.com/apache/httpd/blob/0b61edca6cdda2737aa1d84a4526c5f9d2e23a8c/server/mpm/prefork/prefork.c#L809
[ OK ] No package updates found.
[ -- ] httpd is currently using 35.18 MB of memory.
[ -- ] The smallest apache process is using 1.74 MB of memory
[ -- ] The average apache process is using 1.74 MB of memory
[ -- ] The largest apache process is using 1.74 MB of memory
[ OK ] Going by the average Apache process, Apache can potentially use 445.44 MB RAM:
		Without considering services: 22.38 % of total installed RAM
		Considering extra services: 22.38 % of remaining RAM
[ OK ] Going by the largest Apache process, Apache can potentially use 445.44 MB RAM:
		Without considering services: 22.38 % of total installed RAM
		Considering extra services: 22.38 % of remaining RAM


--------------------------------------------------------------------------------
### GENERAL FINDINGS & RECOMMENDATIONS ###
--------------------------------------------------------------------------------
Apache2buddy.pl report for server: cef6cc345c40 (81.147.138.39):

Settings considered for this report:
[ !! ] *** LOW UPTIME ***.
[ @@ ] The following recommendations may be misleading - apache has been restarted within the last 24 hours.

	Your server's physical RAM:                                   1991 MB
	Remaining Memory after other services considered:             1991 MB
	Apache's MaxRequestWorkers directive:                         256      <--------- Current Setting    
	Apache MPM Model:                                             prefork
	Largest Apache process (by memory):                           1 MB
[ !! ] 	Your MaxRequestWorkers setting is too low.
	Your recommended MaxRequestWorkers setting is between 1029 and 1144.   <------- Acceptable Range (10% of MAX)
	Max potential memory usage:                                   445 MB
	Percentage of TOTAL RAM allocated to Apache:                  22.38  %
	Percentage of REMAINING RAM allocated to Apache:              22.38  %
--------------------------------------------------------------------------------
A log file entry has been made in: /var/log/apache2buddy.log for future reference.

Last 5 entries:

2020/06/16 12:16:51 Uptime: "0d 0h 00m 55s" Model: "Prefork" Memory: "1991 MB" MaxRequestWorkers: "256" Recommended: "1144" Smallest: "1.74 MB" Avg: "1.74 MB" Largest: "1.74 MB" Highest Pct Remaining RAM: "22.38%" (22.38% TOTAL RAM)

[root@cef6cc345c40 /]# 
```

**Comment by richardforth** [comment]
New improved Dockerfile:

```docker
FROM centos:7
RUN yum install -y python3
RUN yum install -y perl
RUN yum install -y which
RUN yum install -y curl
RUN yum install -y net-tools
RUN yum install -y procps
RUN yum install -y httpd
RUN yum install -y httpd-tools
ENTRYPOINT /usr/sbin/httpd -k start && curl -sk https://raw.githubusercontent.com/richardforth/apache2buddy/staging/apache2buddy.pl | perl 
```

to run with SYS_PTRACE cap after build:

```bash
docker run --cap-add SYS_PTRACE <image-id>
```

now starts apache and runs apach2buddy and quits automatically so no more need to hack around in a bash shell.

**Comment by richardforth** [comment]
will create dockerfiles for each flavour we support / test and will add it to Dockerfiles folder in main project, then mark this as solved.

**Comment by richardforth** [comment]
I should just point out to anyone that doesnt understand the purpose of dockerizing apache2buddy - ots for testing only. You will still need to run the apache2buddy script on your actual serversm because apache2buddy will only see what it sees in the containers  - some I'm only using them for testing and hopefully getting changes out faster.

Its also useful to have a known good running script to compare against if you are having issues, that's what Docker is for here, it containerizes it in its own little bubble so it should run all the time, every time - in that container.

So, say you are running Debian 10 and apache2buddy is not working for you, you can build and run run the Debian 10 Dockerfile and check to see why yours is different.

Thats the main idea. You wont be able to tune any existing server using a dockerised version of apache2buddy, it just doesnt work like that, it can only see  whats running inside the container.

**Comment by richardforth** [comment]
This issue is now resolved, the resolving step was this:

```bash
docker run --cap-add SYS_PTRACE <image-id>
```
